rs/messaging/src/routing/stream_handler.rs

```diff
@@ -842,8 +842,20 @@ impl StreamHandlerImpl {
                                 );
                                 return Some((reason, msg));
                             }
+                            RequestOrResponse::Response(response)
+                                if response.deadline != NO_DEADLINE
+                                    && matches!(
+                                        err,
+                                        StateError::CanisterNotFound(_)
+                                            | StateError::CanisterStopped(_)
+                                    ) =>
+                            {
+                                // These are expected edge cases for best effort responses;
+                                // silently drop them.
+                            }
```

Comment:
I wonder this is the right place for this check. We could instead do the one for stopped [here](https://github.com/dfinity/ic/blob/e9ff67e25f5c77ed8e59acd3ba117a4a865e960a/rs/replicated_state/src/canister_state/system_state.rs#L1641) and the one for not found [here](https://github.com/dfinity/ic/blob/e9ff67e25f5c77ed8e59acd3ba117a4a865e960a/rs/replicated_state/src/replicated_state.rs#L797). 

In these places we also have tests already for the edge cases and we could simply extend things. 

What do you think? Would also be curious what @alin-at-dfinity thinks?

Reply:
@alin-at-dfinity on slack:
> I don't have a strong preference regarding where we do it. And as long as we don't miss (or later introduce) yet another edge case, I suppose it's fine to try to be granular about it.

I felt it wasn't a bad idea to put this in one place rather than spread it. You can do it either way, I see no clear reason to do it one way or the other.

But there does seem to be a problem with the metrics and best-effort messages. Currently the incrementation of the metrics [here](https://sourcegraph.com/github.com/dfinity/ic@02b2518bc294230cc5a6c648cd7d1767fb61b54c/-/blob/rs/messaging/src/routing/stream_handler.rs?L812) does not distinguish between successfully inducting and silently dropping since for both cases `Ok(())` is returned. So dropped messages are counted the same as inducted ones.

I ran into this while writing the test I added here. I have removed the check on the metrics for now, but if you try to induct the exact same response twice from the same stream, only one ends up in the state, but both show up in the metrics. A quick and dirty exploration of just changing this to `Ok(true/false)` I did [here](https://github.com/dfinity/ic/pull/3162). It's just an overview of what changes are required. You could also return an error I suppose, probably `NonMatchingResponse` or a new variant. But remember adding a new variant can be more painful than it first appears whereas returning a bool in the case of success seems quite painless.

That is unless I am missing something and this is the behavior we want somehow.

Reply:
> You could also return an error I suppose, probably `NonMatchingResponse` or a new variant. But remember adding a new variant can be more painful than it first appears whereas returning a bool in the case of success seems quite painless.

It's also not precisely an error, strictly speaking. So if we do end up doing something about it, I also believe that a `Response<bool, ...>` is the cleanest approach.

I am wondering however whether this situation can't also occur during the inner loop induction that Execution triggers. It's quite possible that we'd encounter a regular error there, leave the message in place and move on to the next queue; but it's definitely worth checking.

> I wonder this is the right place for this check.

Having had a chance to look at the code a bit, I believe that it would be more appropriate to move the "canister stopped" check into `SystemState::push_input()` (where we otherwise check whether we should be silently dropping the message) and the "canister not found" one into `ReplicatedState::push_input()` (instead of generating the `CanisterNotFound` error). That way, the cause and effect are not separated across two different layers. And it would keep the logic for whether enqueuing a message succeeded or not within `ic_replicated_state`.

I also like your `Result<bool, ...>` change, as that would allow us to simply return `Ok(false)` in those cases instead of the error or a vague `Ok(())`.

Reply:
Moved the checks into the replicated state. Made the test into 3 cases and a common implementation. After this [fix](https://github.com/dfinity/ic/pull/3162) for returning `Result<bool, ..>` goes in, I will add a direct check for the metrics as well.

Reply:
Added tests in the replicated state side as well. I still think the tests with the stream handler have merit because they check the metrics.